File: src/main/java/mnm/mods/util/ILocation.java
package mnm.mods.util;

public interface ILocation {

    int getXPos();

    int getYPos();

    int getWidth();

    int getHeight();

    ILocation asImmutable();

    default int getXWidth() {
        return getXPos() + getWidth();
    }

    default int getYHeight() {
        return getYPos() + getHeight();
    }

    default int getXCenter() {
        return getXPos() + getWidth() / 2;
    }

    default int getYCenter() {
        return getYPos() + getHeight() / 2;
    }

    default Location copy() {
        return Location.copyOf(this);
    }

    default Vec getPoint() {
        return new Vec(getXPos(), getYPos());
    }

    default Dim getSize() {
        return new Dim(this.getWidth(), this.getHeight());
    }

    default boolean contains(ILocation r) {
        return this.getXPos() < r.getXWidth()
                && this.getXWidth() > r.getXPos()
                && this.getYPos() < r.getYHeight()
                && this.getYHeight() > r.getYPos();
    }

    default boolean contains(double x, double y) {
        return this.getXPos() < x
                && this.getYPos() < y
                && this.getXWidth() > x
                && this.getYHeight() > y;
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/util/ILocation.kt
package mnm.mods.tabbychat.util

interface ILocation {

    val xPos: Int
    val yPos: Int
    val width: Int
    val height: Int

    val xWidth: Int get() = xPos + width
    val yHeight: Int get() = yPos + height
    val xCenter: Int get() = xPos + width / 2
    val yCenter: Int get() = yPos + height / 2

    val point: Vec2i get() = Vec2i(xPos, yPos)
    val size: Dim get() = Dim(this.width, this.height)

    fun asImmutable(): ILocation

    fun copy(): Location = Location(this)

    operator fun contains(r: ILocation): Boolean {
        return (this.xPos < r.xWidth
                && this.xWidth > r.xPos
                && this.yPos < r.yHeight
                && this.yHeight > r.yPos)
    }

    fun contains(x: Number, y: Number) = x.toInt() in xPos..xWidth && y.toInt() in yPos..yHeight
}
